### PR TITLE
Don't domain-block URLs that are about to get debounced

### PR DESCRIPTION
**Pull request by pilgrim-brave**
Resolves https://github.com/brave/brave-browser/issues/22437

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by pilgrim-brave** [comment]
This still needs tests, obviously

**Comment by pilgrim-brave** [comment]
Open question: since this will be in the hot path of every top-level navigation, do we want another version of the `Debounce` method that returns on first change? i.e. `bool WouldDebounce(const GURL& original_url)` that simply returns true as soon as a rule changes the URL.

File: components/brave_shields/browser/domain_block_navigation_throttle.h [hunk 1]
```diff
@@ -71,6 +77,7 @@ class DomainBlockNavigationThrottle : public content::NavigationThrottle {
   AdBlockCustomFiltersProvider* ad_block_custom_filters_provider_ = nullptr;
   ephemeral_storage::EphemeralStorageService* ephemeral_storage_service_ =
       nullptr;
+  debounce::DebounceService* debounce_service_ = nullptr;
   HostContentSettingsMap* content_settings_ = nullptr;
```
Comment (goodov) [inline components/brave_shields/browser/domain_block_navigation_throttle.h:81]: `raw_ptr<>` all the things pls

File: components/brave_shields/browser/domain_block_navigation_throttle.cc [hunk 2]
```diff
@@ -122,6 +126,11 @@ DomainBlockNavigationThrottle::WillStartRequest() {
   if (tab_storage->IsProceeding())
     return content::NavigationThrottle::PROCEED;
 
+  // If this URL is about to be debounced, don't show an interstitial.
+  GURL debounced_url;
+  if (debounce_service_->Debounce(request_url, &debounced_url))
```
Comment (goodov) [inline components/brave_shields/browser/domain_block_navigation_throttle.cc:131]: it will crash if `kBraveDebounce` is disabled.

Comment (pilgrim-brave) [reply components/brave_shields/browser/domain_block_navigation_throttle.cc]: fixed in d4cbe8b600

File: components/brave_shields/browser/domain_block_navigation_throttle.h (hunk 1 above)
Comment (pilgrim-brave) [reply components/brave_shields/browser/domain_block_navigation_throttle.h]: fixed in 605806cc51

File: components/brave_shields/browser/domain_block_navigation_throttle.cc [hunk 3]
```diff
@@ -122,6 +126,12 @@ DomainBlockNavigationThrottle::WillStartRequest() {
   if (tab_storage->IsProceeding())
     return content::NavigationThrottle::PROCEED;
 
+  // If this URL is about to be debounced, don't show an interstitial.
+  GURL debounced_url;
+  if (debounce_service_ &&
```
Comment (bridiver) [inline components/brave_shields/browser/domain_block_navigation_throttle.cc:131]: can we accomplish the same thing by just reordering them so debounce is before domain block?

Comment (antonok-edm) [reply components/brave_shields/browser/domain_block_navigation_throttle.cc:136]: @pilgrim-brave were you able to try this approach? any reason it wouldn't work?

**Review by antonok-edm — COMMENTED** [review]
Would be great to see a test too, if possible

**Review by antonok-edm — APPROVED** [review]
beautiful!

**Pull request merged**